### PR TITLE
Update unit tests to be compatible with Spock 2

### PR DESCRIPTION
**Pull request by rieske**
https://github.com/gradle/gradle/issues/15567

This change does not make the tests run on a different version of Spock yet, just makes sure that they are compatible as the first iteration.

This involves:
- removing leftover `@Test` annotations from test classes converted to Spock that are executed with Spock. Those are ignored with current version of Spock, but become an error in Spock 2.
- Fixing unrolled methods that have incorrect `#template` arguments in their names
- Working around the bug with nested assertions: https://github.com/spockframework/spock/issues/1232
- Not verifying stub interactions - mocks used instead
- Fixing field/method name clashes that fail to compile with Spock 2

### CODE REVIEW

File: subprojects/resources/src/test/groovy/org/gradle/internal/resource/BuildOperationFiringExternalResourceDecoratorTest.groovy [hunk 1]
```diff
@@ -141,7 +143,7 @@ class BuildOperationFiringExternalResourceDecoratorTest extends Specification {
         1 * buildOperationExecuter.call(_) >> { CallableBuildOperation op ->
             def operationContextMock = Mock(BuildOperationContext) {
                 1 * setResult(_) >> { ExternalResourceReadBuildOperationType.Result result ->
-                    assert result.bytesRead == TestExternalResource.READ_CONTENT_LENGTH
+                    assertTrue result.bytesRead == TestExternalResource.READ_CONTENT_LENGTH
```
Comment (lptr) [inline subprojects/resources/src/test/groovy/org/gradle/internal/resource/BuildOperationFiringExternalResourceDecoratorTest.groovy:146]: Why is this change needed?

**Review by lptr — COMMENTED** [review]
I left a few comments.

It would be nice to have a summary published about the changes we have to keep in mind using Spock 2 on Slack or something. I'm especially surprised by the changes to exception variable names and the use of `assertTrue` instead of `assert`.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/project/DefaultProjectStateRegistryTest.groovy [hunk 2]
```diff
@@ -412,16 +414,16 @@ class DefaultProjectStateRegistryTest extends ConcurrentSpec {
...
-                        assert state1.hasMutableState()
-                        assert !state2.hasMutableState()
+                        assertTrue state1.hasMutableState()
+                        assertTrue !state2.hasMutableState()
                     }
                     state2.applyToMutableState {
-                        assert state1.hasMutableState()
-                        assert state2.hasMutableState()
+                        assertTrue state1.hasMutableState()
+                        assertTrue state2.hasMutableState()
```
[the remark this answers is not in the document]
Comment (rieske) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/project/DefaultProjectStateRegistryTest.groovy:426]: This is a workaround for this bug where asserts nested in closures break with static compile: spockframework/spock#1232

File: subprojects/resources/src/test/groovy/org/gradle/internal/resource/BuildOperationFiringExternalResourceDecoratorTest.groovy (hunk 1 above)
Comment (rieske) [reply subprojects/resources/src/test/groovy/org/gradle/internal/resource/BuildOperationFiringExternalResourceDecoratorTest.groovy:146]: Same, a workaround to spockframework/spock#1232

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy [hunk 3]
```diff
@@ -109,7 +110,7 @@ class ExecuteActionsTaskExecuterTest extends Specification {
         getOutputFilesProducedByWork() >> ImmutableSortedMap.of()
     }
     def validationContext = new DefaultWorkValidationContext()
-    def executionContext = Stub(TaskExecutionContext)
+    def executionContext = Mock(TaskExecutionContext)
```
[the remark this answers is not in the document]
Comment (rieske) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuterTest.groovy:113]:
```
Stub 'executionContext' matches the following required interaction:

0 * executionContext._   (0 invocations)

Remove the cardinality (e.g. '1 *'), or turn the stub into a mock.
```

File: subprojects/build-operations/src/test/groovy/org/gradle/internal/operations/DefaultBuildOperationRunnerTest.groovy [hunk 4]
```diff
@@ -182,8 +182,8 @@ class DefaultBuildOperationRunnerTest extends Specification {
 
         where:
         description        | expectedException             | fail
-        "thrown exception" | new RuntimeException("Error") | { BuildOperationContext context, Exception expectedException -> throw expectedException }
-        "context"          | null                          | { BuildOperationContext context, Exception expectedException -> context.failed(new RuntimeException("Error")) }
+        "thrown exception" | new RuntimeException("Error") | { BuildOperationContext context, Exception e -> throw e }
+        "context"          | null                          | { BuildOperationContext context, Exception e -> context.failed(new RuntimeException("Error")) }
```
[the remark this answers is not in the document]
Comment (rieske) [reply subprojects/build-operations/src/test/groovy/org/gradle/internal/operations/DefaultBuildOperationRunnerTest.groovy:186]: I'm not sure what exactly changed in Spock and/or whether this is related to StaticCompile, but the compilation fails when using spock2 otherwise: https://ge.gradle.org/s/zy4ecethfbbu4/console-log?anchor=384

File: subprojects/messaging/src/test/groovy/org/gradle/internal/remote/internal/inet/TcpConnectorTest.groovy [hunk 5]
```diff
@@ -207,9 +207,9 @@ class TcpConnectorTest extends ConcurrentSpec {
 
         when:
         def acceptor = incomingConnector.accept({ ConnectCompletion event ->
-            def connection = event.create(serializer)
-            connection.dispatch("bye")
-            connection.stop()
+            def conn = event.create(serializer)
+            conn.dispatch("bye")
+            conn.stop()
```
[the remark this answers is not in the document]
Comment (rieske) [reply subprojects/messaging/src/test/groovy/org/gradle/internal/remote/internal/inet/TcpConnectorTest.groovy:212]: https://ge.gradle.org/s/c6fc4hdngbqzk/console-log?anchor=595
This again, is a workaround to make it compile, not about readability.
It is weird, but for some reason it looks like the static compilation treats it as a name clash when a variable is declared both inside and outside of a closure.
I choose to change/shorten the name of the one with the narrowest scope.

File: subprojects/resources-http/src/test/groovy/org/gradle/internal/resource/transport/http/HttpResponseResourceTest.groovy [hunk 6]
```diff
@@ -45,10 +45,10 @@ class HttpResponseResourceTest extends AbstractHttpClientTest {
         setup:
         1 * response.entity >> Mock(HttpEntity)
         when:
-        def resource = resource();
-        resource.openStream();
+        def r = resource()
+        r.openStream()
         and:
-        resource.openStream()
+        r.openStream()
```
[the remark this answers is not in the document]
Comment (rieske) [reply subprojects/resources-http/src/test/groovy/org/gradle/internal/resource/transport/http/HttpResponseResourceTest.groovy]: This time there's a runtime failure `Expected exception of type 'java.io.IOException', but got 'java.lang.NullPointerException'` - looks like Spock2 fails when the variable and a method in the test share the name. I don't like it either, but it works.

And just figured there is a better workaround - the method in this case can be prefixed with `this.` and then it works.

File: subprojects/logging/src/test/groovy/org/gradle/internal/deprecation/LoggingDeprecatedFeatureHandlerTest.groovy [hunk 7]
```diff
@@ -387,21 +387,12 @@ class LoggingDeprecatedFeatureHandlerTest extends Specification {
...
-        then:
-        1 * buildOperationListener.progress(_, _) >> { progressFired(it[1], 'feature2') }
-
-        when:
         handler.featureUsed(deprecatedFeatureUsage('feature2'))
 
         then:
-        1 * buildOperationListener.progress(_, _) >> { progressFired(it[1], 'feature2') }
+        1 * buildOperationListener.progress(_, _) >> { progressFired(it[1], 'feature1') }
+        2 * buildOperationListener.progress(_, _) >> { progressFired(it[1], 'feature2') }
```
[the remark this answers is not in the document]
Comment (rieske) [reply subprojects/logging/src/test/groovy/org/gradle/internal/deprecation/LoggingDeprecatedFeatureHandlerTest.groovy]: It was asserting the same thing in order.

I managed to make the test as explicit as it was again by removing multiple when/then/when/thens.
Apparently something goes terribly upside down when this is combined with mock assertions with closures:
```
Caused by: java.lang.VerifyError: Expecting a stackmap frame at branch target 20
Exception Details:
  Location:
    org/gradle/internal/deprecation/LoggingDeprecatedFeatureHandlerTest.$getCallSiteArray()[Lorg/codehaus/groovy/runtime/callsite/CallSite; @3: ifnull
  Reason:
    Expected stackmap frame at this location.
  Bytecode:
    0000000: b203 fec6 0011 b203 feb6 0403 c003 f959
    0000010: 4bc7 0012 b804 054b bb04 0059 2ab7 0407
    0000020: b303 fe2a b404 0bb0                    
```

**Review by jjohannes — APPROVED** [review]
LGTM
